python/seldon_core/batch_processor.py

```diff
@@ -316,23 +320,102 @@ def _send_batch_predict_multi_request(
+                arrays = [np.array(d["data"]["ndarray"]) for d in loaded_data]
+                if not all_equal(names_list):
+                    raise ValueError("All names in mini-batch must be the same.")
+                else:
+                    names = names_list[0]
+                for arr in arrays:
+                    if arr.shape[0] != 1:
+                        raise ValueError(
+                            "When using mini-batching each row should contain single instance."
+                        )
+                ndarray = np.concatenate(arrays)
+                raw_data = {
+                    "data": {"names": names_list[0], "ndarray": ndarray.tolist()},
+                    "meta": {"tags": predict_kwargs["meta"]},
+                }
+                predict_kwargs["raw_data"] = raw_data
+            # If-block for tensor case
+            elif "tensor" in first_input["data"]:
+                payload_type = "tensor"
+                names_list = [d["data"]["names"] for d in loaded_data]
+                tensor_shapes = [d["data"]["tensor"]["shape"] for d in loaded_data]
+                tensor_values = [d["data"]["tensor"]["values"] for d in loaded_data]
+
+                if not all_equal(names_list):
+                    raise ValueError("All names in mini-batch must be the same.")
+
+                dim_0 = 0
+                dim_1 = tensor_shapes[0][1]
+                for shape in tensor_shapes:
+                    if shape[0] != 1:
+                        raise ValueError(
+                            "When using mini-batching each row should contain single instance."
+                        )
+                    dim_0 += shape[0]
+                    if dim_1 != shape[1]:
+                        raise ValueError(
+                            "All instances in mini-batch must have same number of features."
+                        )
+                values = sum(tensor_values, [])
+                shape = [dim_0, dim_1]
+                raw_data = {
+                    "data": {
+                        "names": names_list[0],
+                        "tensor": {"shape": shape, "values": values},
+                    },
+                    "meta": {"tags": predict_kwargs["meta"]},
+                }
+                predict_kwargs["raw_data"] = raw_data
+        else:
+            # Initialise concatenated array for data
+            arrays = [np.array(arr) for arr in loaded_data]
+            for arr in arrays:
+                if arr.shape[0] != 1:
+                    raise ValueError(
+                        "When using mini-batching each row should contain single instance."
+                    )
+            concat = np.concatenate(arrays)
+            predict_kwargs["data"] = concat
+        logger.info(f"calling sc.predict with {predict_kwargs}")
+    except Exception as e:
+        error_resp = {
+            "status": {"info": "FAILURE", "reason": str(e), "status": 1},
+            "meta": tags,
```

Comment:
Would batch id not be part of this meta?

Reply:
Actually, it seems it would be only `batch_id`
```
    tags = {
        "batch_id": batch_id,
    }
```

This goes along what was happening in other `except Exception as e` block bellow

Reply:
Hi Rafal,  In my local code, before I found the error message is empty when there is 503 and 400 errors.  So I added something like the belows:
in the beginning of the method:
```
msg = ""
```,
in the requests part:
```
                seldon_payload = sc.predict(**predict_kwargs)
                msg = seldon_payload.msg
                logger.error(f"Exception: {e}, retries {i}, message:{msg}, request data:{data_np}")
```
At that time the request data "data_np" includes only data part or the "feature:value" depends on the input definition.   I find these information are quite useful for debugging then batch_id. Batch_id is generated  and hard to get the request input from that field if we look at the input file.

Reply:
Hi Rafal,  In my local code, before I found the error message is empty when there is 503 and 400 errors.  So I added something like the belows:
in the beginning of the method:
```
msg = ""
```,
in the requests part:
```
                seldon_payload = sc.predict(**predict_kwargs)
                msg = seldon_payload.msg
                logger.error(f"Exception: {e}, retries {i}, message:{msg}, request data:{data_np}")
```
At that time the request data "data_np" includes only data part or the "feature:value" depends on the input definition.   I find these information are quite useful for debugging then batch_id. Batch_id is generated  and hard to get the request input from that field if we look at the input file.

Reply:
We could print `batch_index` which basically corresponds to the line number

Reply:
With that exact code snippet I am afraid that for
```
seldon_payload = sc.predict(**predict_kwargs)
msg = seldon_payload.msg
```
it may catch exception on first line and the `msg` printed may be not accurate

Reply:
@yaliqin I added printing of batch indices (which basically corresponds to the number of line in the input file) - le me know if this will help in the debugging of issues (
also if log level will be setup to DEBUG it will now log predict_kwargs)

Reply:
Also, now in the output file there will be single line of output for each line of input like with BATCH_SIZE=1 option

Reply:
Thanks for the modification. We currently use the batch processor in the pipeline. During the pipeline running I read this step log file for debugging. The line number is useful for debugging.